### PR TITLE
Namespaced notification feeds

### PR DESCRIPTION
**Pull request by jguery**
This PR contains APIs changes - although only internal APIs.
Namely, we used to have:
 - PUT/DELETE/GET v3/feeds/{feed-id} with feed-id = [namespace].[category].[name]

We now have:
 - PUT/DELETE/GET v3/namespaces/{namespace}/feeds/categories/{category}/names/{name}

Build: https://builds.cask.co/browse/CDAP-DUT682

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NotificationFeedHttpHandler.java [hunk 1]
```diff
@@ -63,13 +63,19 @@ public NotificationFeedHttpHandler(Authenticator authenticator, NotificationFeed
   }
 
   @PUT
-  @Path("/feeds/{id}")
-  public void createFeed(HttpRequest request, HttpResponder responder, @PathParam("id") String id) {
+  @Path("/feeds/categories/{feed-category}/names/{feed-name}")
+  public void createFeed(HttpRequest request, HttpResponder responder,
+                         @PathParam("namespace-id") final String namespaceId,
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NotificationFeedHttpHandler.java]: Params don't need to be ``final``?

Comment (jguery) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NotificationFeedHttpHandler.java]: They don't need to, but that's the convention I've found in other handlers though.

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NotificationFeedHttpHandler.java]: There is no such convention. Shouldn't have final.

Comment (jguery) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NotificationFeedHttpHandler.java]: Ok.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/AbstractStreamFileAdmin.java [hunk 2]
```diff
@@ -314,15 +314,15 @@ public void create(String name, @Nullable Properties props) throws Exception {
   }
 
   /**
-   * Create the public {@link NotificationFeed}s that concerns the stream with configuration {@code config}.
+   * Create the public {@link Id.NotificationFeed}s that concerns the stream with configuration {@code config}.
    *
    * @param config config of the stream to create feeds for
    */
   private void createStreamFeeds(StreamConfig config) {
     // TODO use accountID as namespace?
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/AbstractStreamFileAdmin.java]: do we need this todo?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/AbstractStreamFileAdmin.java [hunk 3]
```diff
@@ -314,15 +314,15 @@ public void create(String name, @Nullable Properties props) throws Exception {
...
     try {
-      NotificationFeed streamFeed = new NotificationFeed.Builder()
-        .setNamespace(Constants.DEFAULT_NAMESPACE)
+      Id.NotificationFeed streamFeed = new Id.NotificationFeed.Builder()
+        .setNamespaceId(Constants.DEFAULT_NAMESPACE)
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/AbstractStreamFileAdmin.java]: So we're not really namespacing it at this point right? everything goes into *default* namespace?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NotificationFeedHttpHandler.java [hunk 4]
```diff
@@ -63,13 +63,19 @@ public NotificationFeedHttpHandler(Authenticator authenticator, NotificationFeed
...
+      Id.NotificationFeed combinedFeed;
       try {
-        NotificationFeed feed = parseBody(request, NotificationFeed.class);
-        combinedFeed = new NotificationFeed.Builder(NotificationFeed.fromId(id))
+        Id.NotificationFeed feed = parseBody(request, Id.NotificationFeed.class);
+        combinedFeed = new Id.NotificationFeed.Builder()
+          .setNamespaceId(namespaceId)
+          .setCategory(category)
+          .setName(name)
           .setDescription(feed == null ? null : feed.getDescription())
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NotificationFeedHttpHandler.java]: What do ``null``s get persisted as? Should we make it an empty string instead?

File: cdap-notifications-api/src/main/java/co/cask/cdap/notifications/feeds/NotificationFeedManager.java [hunk 5]
```diff
@@ -31,31 +33,32 @@
    * @throws NotificationFeedException if the feed has an incorrect structure;
    * for example, if it is missing a name, a namespace or a category
    */
-  boolean createFeed(NotificationFeed feed) throws NotificationFeedException;
+  boolean createFeed(Id.NotificationFeed feed) throws NotificationFeedException;
```
Comment (bdmogal) [inline cdap-notifications-api/src/main/java/co/cask/cdap/notifications/feeds/NotificationFeedManager.java]: Is it better to have the ``createFeed`` throw an AlreadyExistsException? An alternative is to use the [ConcurrentHashMap#putIfAbsent](http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentHashMap.html#putIfAbsent\(K,%20V\)) signature. 

A create returning false if was already present seems a little strange.

File: cdap-notifications-api/src/main/java/co/cask/cdap/notifications/feeds/NotificationFeedManager.java [hunk 6]
```diff
@@ -31,31 +33,32 @@
...
    * Deletes the {@code feed} from the manager store. To determine if the feed exists,
-   * the {@link NotificationFeed#getId} method is used.
+   * the {@link Id.NotificationFeed#getId} method is used.
    *
-   * @param feed the {@link NotificationFeed} to delete
+   * @param feed the {@link Id.NotificationFeed} to delete
    * @throws NotificationFeedNotFoundException if the feed does not exist
    */
-  void deleteFeed(NotificationFeed feed) throws NotificationFeedException;
+  void deleteFeed(Id.NotificationFeed feed) throws NotificationFeedException;
```
Comment (bdmogal) [inline cdap-notifications-api/src/main/java/co/cask/cdap/notifications/feeds/NotificationFeedManager.java]: Should this also throw a ``NotFoundException``.

Maybe unrelated to this PR, but something to think about.

File: cdap-notifications-api/src/main/java/co/cask/cdap/notifications/feeds/NotificationFeedNotFoundException.java [hunk 7]
```diff
@@ -16,8 +16,10 @@
 
 package co.cask.cdap.notifications.feeds;
 
+import co.cask.cdap.proto.Id;
+
 /**
- * Exception thrown when a {@link NotificationFeed} object is not found.
+ * Exception thrown when a {@link Id.NotificationFeed} object is not found.
  */
 public class NotificationFeedNotFoundException extends NotificationFeedException {
```
Comment (bdmogal) [inline cdap-notifications-api/src/main/java/co/cask/cdap/notifications/feeds/NotificationFeedNotFoundException.java]: We have ``co.cask.cdap.common.NotFoundException``. Could you use that instead? For a later PR maybe.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java [hunk 8]
```diff
@@ -200,4 +201,154 @@ public String toString() {
...
+                                  "Namespace, category or name has a wrong format.");
+
+      this.namespace = Namespace.from(namespace);
+      this.category = category;
+      this.name = name;
+      this.id = String.format("%s.%s.%s", namespace, category, name);
+      this.description = description;
+    }
+
+    private boolean isId(String name) {
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Any way to make this method available to all ``Id`` classes? Maybe you can expose it as part of this PR, we can use it in the other ``Id`` classes later?

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java [hunk 9]
```diff
@@ -200,4 +201,154 @@ public String toString() {
...
+    }
+
+    public String getDescription() {
+      return description;
+    }
+
+    /**
+     * Builder used to build {@link NotificationFeed}.
+     */
+    public static final class Builder {
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Could ``Id.NotificationFeed`` grow to contain a lot more fields? Just thinking that its construction (via ``Builder``) is not consistent with the other ``Id`` classes (which expose the constructor directly).

File: cdap-notifications-api/src/main/java/co/cask/cdap/notifications/feeds/NotificationFeedManager.java (hunk 5 above)
Comment (jguery) [inline cdap-notifications-api/src/main/java/co/cask/cdap/notifications/feeds/NotificationFeedManager.java]: It doesn't seem strange to me at all: you want to perform an action, and it returns false if that action did not work. Then the javadoc tells you in which case it did not work. It kind of is the same as the ``Collection.add`` method signature (http://docs.oracle.com/javase/7/docs/api/java/util/Collection.html#add(E)). It returns true if the collection as changed, just like we do here, and false if not, again like we do here.

I feel like a lot of classes follow various contracts, and in the end not any one is better than the other.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java [hunk 10]
```diff
@@ -200,4 +201,154 @@ public String toString() {
...
+       * @return a {@link NotificationFeed} object containing all the fields set in the builder.
+       * @throws IllegalArgumentException if the namespaceId, category or name is invalid.
+       */
+      public NotificationFeed build() {
+        return new NotificationFeed(namespaceId, category, name, description);
+      }
+    }
+
+    @Override
+    public String toString() {
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: I can see that the ``id`` here contains namespace, category and name, but better to use all fields in ``toStringHelper()`` so the formatting is consistent?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/AbstractStreamFileAdmin.java (hunk 3 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/AbstractStreamFileAdmin.java]: I think this should be changed once stream are being namespaced. At this point we don't have the namespace information from the stream we try to create a feed for.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java [hunk 11]
```diff
@@ -200,4 +201,154 @@ public String toString() {
...
+
+  /**
+   * Represents ID of a Notification feed.
+   */
+  public static class NotificationFeed {
+
+    /**
+     * Id is composed of: [namespaceId].[category].[name]
+     */
+    private final String id;
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Can this work without the ``id`` member, which contains the other fields?

Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Accordingly please update ``toString()``, ``equals()`` and ``hashCode()``

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NotificationFeedHttpHandler.java (hunk 4 above)
Comment (jguery) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NotificationFeedHttpHandler.java]: Gson will simply not include the field in the serialized string, when this one is ``null``. Then when deserializing, it will set the field as null, as expected.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/AbstractStreamFileAdmin.java (hunk 2 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/AbstractStreamFileAdmin.java]: No you're right. It will anyway be covered once we namespace streams.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/AbstractStreamFileAdmin.java (hunk 3 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/AbstractStreamFileAdmin.java]: Could we miss this while namespacing Streams though? Unsure if @anwar6953 is factoring changes here while namespacing streams. Please co-ordinate.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java (hunk 8 above)
Comment (jguery) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Alright.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java (hunk 9 above)
Comment (jguery) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Yes we could have more metadata later.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java (hunk 10 above)
Comment (jguery) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Yes, will do.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java (hunk 11 above)
Comment (jguery) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Ok.

**Comment by jguery** [comment]
@bdmogal Addressed your comments, please have another look. Also, what should be done to check existence of a namespace when creating a feed?

File: cdap-notifications-api/src/main/java/co/cask/cdap/notifications/feeds/NotificationFeedNotFoundException.java (hunk 7 above)
Comment (bdmogal) [inline cdap-notifications-api/src/main/java/co/cask/cdap/notifications/feeds/NotificationFeedNotFoundException.java]: Any thoughts about this? Will this change increase the size of this PR a lot?

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java [hunk 12]
```diff
@@ -200,4 +209,144 @@ public String toString() {
...
+      }
+      return new NotificationFeed(idParts[0], idParts[1], idParts[2], "");
+    }
+
+    private NotificationFeed(String namespace, String category, String name, String description) {
+      Preconditions.checkArgument(namespace != null && !namespace.isEmpty(),
+                                  "Namespace value cannot be null or empty.");
+      Preconditions.checkArgument(category != null && !category.isEmpty(), "Category value cannot be null or empty.");
+      Preconditions.checkArgument(name != null && !name.isEmpty(), "Name value cannot be null or empty.");
+      Preconditions.checkArgument(isId(namespace) && isId(category) && isId(name),
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Should we throw a checked exception if these are not valid identifiers? This message should somehow be communicated to clients as a 400 BAD_REQUEST, not a 500 INTERNAL_SERVER_ERROR, no? Difficult to catch runtime exceptions and send BAD_REQUEST.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java [hunk 13]
```diff
@@ -200,4 +209,144 @@ public String toString() {
...
+      this.category = category;
+      this.name = name;
+      this.description = description;
+    }
+
+    public String getCategory() {
+      return category;
+    }
+
+    public String getId() {
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Do we still use the String id in notifications?

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java (hunk 12 above)
Comment (jguery) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: We do catch the ``IllegalArgumentException`` thrown by those ``Preconditions.checkArgument`` calls: in ``NotificationFeedHttpHandler``, there is a try-catch around the creation of a ``NotificationFeed``, and we send 400 in the catch.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java (hunk 13 above)
Comment (jguery) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Yes, we still use the ``id`` as the key to store a ``NotificationFeed``, in the metastore.

Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Ok.

File: cdap-notifications-api/src/main/java/co/cask/cdap/notifications/feeds/NotificationFeedNotFoundException.java (hunk 7 above)
Comment (jguery) [inline cdap-notifications-api/src/main/java/co/cask/cdap/notifications/feeds/NotificationFeedNotFoundException.java]: It would change a few things here and there yes. I could do that in another PR, this is not related to the work here.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java (hunk 12 above)
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: Yeah, the concern I have is, since we are exposing the ``isId`` as sharable between all ``Id``s, it would be good if it throws an ``InvalidIdException`` or something similar, which all constructors in the ``Id`` subclasses would propagate, so anytime you create an ``Id.Application`` or ``Id.Program`` or anything else,  it clearly indicates to clients that they need to handle that case.

File: cdap-notifications-api/src/main/java/co/cask/cdap/notifications/feeds/NotificationFeedNotFoundException.java (hunk 7 above)
Comment (bdmogal) [inline cdap-notifications-api/src/main/java/co/cask/cdap/notifications/feeds/NotificationFeedNotFoundException.java]: Sounds good.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java (hunk 12 above)
Comment (jguery) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: I would say, let's limit the scope of this PR to namespace related stuff. I can always open a small PR for that change.

Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: sure.

**Comment by bdmogal** [comment]
LGTM. Please open follow up PRs for ``NotFoundException`` and especially for the actual (non-default) namespace integration.

**Comment by anwar6953** [comment]
Looking at the NotificationFeed changes in FileStreamAdmin, I should note that the FileStreamAdmin is not aware of what namespace it is operating under (simply put, the fields in the constructor such as streamBaseLocation will be namespace-specific, but otherwise the actions it takes will not be considering namespace).

Because of that, we should probably make the NotificationFeedManager similar (since it is used within a FileStreamAdmin).

**Comment by jguery** [comment]
@anwar6953 if that is the case, do you also have one ``StreamCoordinatorClient`` per namespace? How does that work?

**Comment by anwar6953** [comment]
Yeah - I'm going to have to figure out some way to do that. Either a delegating class that namespaces, or refactor the StreamCoordinatorClient.

**Comment by jguery** [comment]
For each component touched by namespaces we seem to have a different way to deal with them... I'd like to understand the bigger picture here. How can the ``NotificationFeedManager`` be namespace agnostic if, the store in which it retrieves/stores feeds needs to be given a namespace ID? 

I still think we should have one store per namespace, even if it doesn't translate to one dataset per store, but only by having one dataset, and each store just prefixes all the row keys it writes with a namespace ID.

**Comment by anwar6953** [comment]
You're right that each component deals with namespaces differently. But at least going forward, we should do it the way that the component doesn't know about the namespace entity. Simply the resources it is given are already namespaced (such as the store).

This will heavily decrease the work required in the future, when implementing accounts within a namespace (different accounts for developer, ops, admin, etc).

**Comment by jguery** [comment]
After further discussion, it appears that we will have one ``StreamAdmin`` for all namespaces, and its methods will be passed a ``Id.Stream`` containing the namespace.
This PR can now be merged.

**Comment by anwar6953** [comment]
Right. Agreed that StreamAdmin will be aware of the namespace of any streams it acts on.
